File: stateful-server/sessions.js
/**
 * simple session management object
 * Uses Set-Cookie and Cookie HTTP Headers to maintain session at the server
 * 
 * Created for MDW CTU Course, 2011 
 * 
 * @author
 * Tomas Vitvar, http://vitvar.com
 * 
 * @License
 * MIT License
 */

var sessions = {
    // property to hold all sessions from all clients
    sessions : {},

    // returns the session data based on the information in the cookie
    getSession : function(req) {
        var sid;

        // get the session id from the cookie
        if (req.headers.cookie && (p = req.headers.cookie.match(".*session-id=([a-zA-Z0-9]+).*")))
            sid = p[1];

        if (!sid || !this.sessions[sid])
            // create the session id md5 hash; maximum of 1000 connections 
            // from a single IP
            sid = require("crypto").createHash('md5').update(
                req.connection.remoteAddress + 
                    (Math.floor(Math.random()*999))).digest("hex");

        // return the session data; 
        // create the empty object if data does not exist
        return this.sessions[sid] || 
            (this.sessions[sid] = { id : sid, data : {},
                header: { "Set-Cookie" : "session-id=" + sid + "; Max-Age=3600" } });
    }
};

exports.sessions = sessions;

File: protocols/src/com/vitvar/ctu/mdw/Sessions.java
package com.vitvar.ctu.mdw;

import java.security.MessageDigest;
import java.util.Hashtable;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javax.servlet.http.HttpServletRequest;

import com.vitvar.ctu.mdw.commons.Utils;

/** Class for the session management that uses HTTP cookies **/
public class Sessions<E> {
	
	// storage for the session data;
	private Hashtable<String, E> sessions = new Hashtable<String, E>();
	
	/** Returns session id based on the information in the http request **/
	public String getSessionID(HttpServletRequest request) throws Exception {
		String sid = null;
		
		// extract the session id from the cookie
		if (request.getHeader("cookie") != null) {
			Pattern p = Pattern.compile(".*session-id=([a-zA-Z0-9]+).*");
			Matcher m = p.matcher(request.getHeader("cookie"));
			if (m.matches()) sid = m.group(1);
		}
		
		// create the session id md5 hash; max 1000 connections per IP 
		if (sid == null || sessions.get(sid) == null) {
			MessageDigest md = MessageDigest.getInstance("MD5");
			md.update(new String(request.getRemoteAddr() + 
				Math.floor(Math.random()*999)).getBytes());
			sid = Utils.toHexString(md.digest());
		}
		return sid;
	}
	
	/** returns the session data based on a session id */
	public E getData(String sid) {
		return sessions.get(sid);
	}
	
	/** sets the session data for a specified session id */
	public void setData(String sid, E d) {
		sessions.put(sid, d);
	}
}
